### PR TITLE
Improve performance and simplify map equal and distinct from

### PR DESCRIPTION
**Pull request by gerlou**
Map Equal and Distinct From no longer implemented by making Java maps

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java [hunk 1]
```diff
@@ -41,29 +40,23 @@ public static Boolean genericEqual(
...
-        for (Map.Entry<KeyWrapper, Integer> entry : wrappedRightMap.entrySet()) {
-            KeyWrapper key = entry.getKey();
-            Integer leftValuePosition = wrappedLeftMap.get(key);
-            if (leftValuePosition == null) {
+        SingleMapBlock mapLeftBlock = (SingleMapBlock) leftMapBlock;
+        SingleMapBlock mapRightBlock = (SingleMapBlock) rightMapBlock;
+
+        for (int position = 0; position < leftMapBlock.getPositionCount(); position += 2) {
+            Object key = readNativeValue(keyType, mapLeftBlock, position);
+            int leftPosition = mapLeftBlock.seekKey(key);
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java]: `int leftPosition = position + 1`

Map block in Presto is laid out as `k1, v1, k2, v2, k3, v3, ...`. This is part of the contract of the map block. You can safely assume that. Therefore, it is unnecessary to call seekKey on the blocks.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapDistinctFromOperator.java [hunk 2]
```diff
@@ -59,6 +59,7 @@ public static boolean isDistinctFrom(
         if (leftMapNull) {
             return false;
         }
+
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapDistinctFromOperator.java]: unrelated change

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java [hunk 3]
```diff
@@ -41,29 +40,23 @@ public static Boolean genericEqual(
...
+            if (rightPosition == -1) {
                 return false;
             }
 
             try {
-                Boolean result = predicate.equals(leftValuePosition, entry.getValue());
+                Boolean result = predicate.equals(leftPosition, rightPosition);
                 if (result == null) {
                     return null;
                 }
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java]: `KeyWrapper` class is now unused, should be removed

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java [hunk 4]
```diff
@@ -41,29 +40,23 @@ public static Boolean genericEqual(
             Block rightMapBlock,
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java]: `keyEqualsFunction, keyHashcodeFunction` method params are now unused. Remove?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java [hunk 5]
```diff
@@ -34,36 +31,28 @@ Boolean equals(int leftMapIndex, int rightMapIndex)
...
-        if (wrappedLeftMap.size() != wrappedRightMap.size()) {
+        if (leftMapBlock.getPositionCount() != rightMapBlock.getPositionCount()) {
             return false;
         }
 
-        for (Map.Entry<KeyWrapper, Integer> entry : wrappedRightMap.entrySet()) {
-            KeyWrapper key = entry.getKey();
-            Integer leftValuePosition = wrappedLeftMap.get(key);
-            if (leftValuePosition == null) {
+        SingleMapBlock mapLeftBlock = (SingleMapBlock) leftMapBlock;
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java]: nit: I would call it `leftSingleMapBlock`, and make the argument parameter name to be `leftBlock`. ditto for rightBlock

File: presto-main/src/test/java/com/facebook/presto/type/TestMapOperators.java [hunk 6]
```diff
@@ -702,6 +702,8 @@ public void testDistinctFrom()
         assertFunction("MAP(ARRAY[1], ARRAY[NULL]) IS DISTINCT FROM MAP(ARRAY[1], ARRAY[2])", BOOLEAN, true);
         assertFunction("MAP(ARRAY[1], ARRAY[2]) IS DISTINCT FROM MAP(ARRAY[1], ARRAY[NULL])", BOOLEAN, true);
         assertFunction("MAP(ARRAY[1], ARRAY[NULL]) IS DISTINCT FROM MAP(ARRAY[1], ARRAY[NULL])", BOOLEAN, false);
+        assertFunction("MAP(ARRAY[1], ARRAY[NULL]) IS DISTINCT FROM MAP(ARRAY[1], ARRAY[0])", BOOLEAN, true);
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/type/TestMapOperators.java]: Doesn't this test case serve the same purpose with the one on line 702? @haozhun
